### PR TITLE
feat: HTTP headers in Error type

### PR DESCRIPTION
**Pull request by karel-rehor**
## Proposed Changes

- Add `http.Header` as `Header` property to `api/http/Error` type to expose response headers for analysis and debugging.
- leverage this in `api/http/service,go`.
- add selected headers to error message log.
- add new tests covering the new property.
- update example to show how the new property can be used. 

## Checklist

- [x] CHANGELOG.md updated
- [x] Rebased/mergeable
- [x] A test has been added if appropriate
- [x] Tests pass
- [x] Commit messages are in [semantic format](https://seesparkbox.com/foundry/semantic_commit_messages)
- [x] Sign [CLA](https://www.influxdata.com/legal/cla/) (if not already signed)

### CODE REVIEW

File: internal/write/service_test.go [hunk 1]
```diff
@@ -337,10 +337,13 @@ func TestMaxRetryTime(t *testing.T) {
 	b := NewBatch("2\n", exp)
 	// First batch will  be checked against maxRetryTime and it will expire. New batch will fail and it will added to retry queue
 	err = srv.HandleWrite(ctx, b)
+	//fmt.Printf("DEBUG err %v\n", err)
+	//fmt.Printf("DEBUG err %v\n", err)
```
Comment (vlastahajek) [inline internal/write/service_test.go:341]: Remove commented code

File: internal/write/service_test.go [hunk 2]
```diff
@@ -337,10 +337,13 @@ func TestMaxRetryTime(t *testing.T) {
...
 	require.NotNil(t, err)
 	// 1st Batch expires and writing 2nd trows error
 	assert.Equal(t, "write failed (attempts 1): Unexpected status code 429", err.Error())
 	assert.Equal(t, 1, srv.retryQueue.list.Len())
+	// fmt.Printf("DEBUG Header len: %d\n", len(err.(*http.Error).Header))
```
Comment (vlastahajek) [inline internal/write/service_test.go:346]: Remove commented code

File: internal/write/service.go [hunk 3]
```diff
@@ -196,9 +197,30 @@ func (w *Service) HandleWrite(ctx context.Context, batch *Batch) error {
...
+					}
+					return &http2.Error{
+						StatusCode: int(perror.StatusCode),
+						Code:       perror.Code,
+						Message: fmt.Errorf(
+							"write failed (attempts %d): %w", batchToWrite.RetryAttempts, perror,
+						).Error(),
+						Err:        perror.Err,
+						RetryAfter: perror.RetryAfter,
+						Header:     perror.Header,
```
Comment (vlastahajek) [inline internal/write/service.go:222]: It doesn't look good from the API perspective to return `http.Error` from the write service. And the more serious problem is that the final message looks silly: `"400 Bad Request: write failed (attempts 0): 400 Bad Request: { \"code\": \"bad request\", \"message\": \"test header\"`.

The solution could be to create a write error that would wrap the previous:
```go
package write

type Error struct {
	origin  error
	Message string
}

func NewError(err error, message string) *Error {
	return &Error{
		origin:  err,
		Message: message,
	}
}

func (e *Error) Error() string {
	return fmt.Sprintf("%s: %s", e.Message, e.origin)
}

func (e *Error) Unwrap() error {
	return e.origin
}
```
Then this code would be:
```go
return write.NewError(perror, fmt.Sprintf("write failed (attempts %d)", batchToWrite.RetryAttempts))
```

**Review by vlastahajek — CHANGES_REQUESTED** [review]

File: api/write_test.go [hunk 4]
```diff
@@ -265,3 +268,88 @@ func TestFlushWithRetries(t *testing.T) {
...
+	points := test.GenPoints(15)
+	for i := 0; i < 15; i++ {
+		writeAPI.WritePoint(points[i])
+	}
+	writeAPI.waitForFlushing()
+	wg.Wait()
+	assert.Equal(t, calls, 3)
+}
+
+func TestWriteErrorHeaderToString(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply api/write_test.go]: refactored to error_test.go

File: internal/write/service.go [hunk 5]
```diff
@@ -164,6 +164,7 @@ func (w *Service) HandleWrite(ctx context.Context, batch *Batch) error {
 		if batchToWrite != nil {
 			perror := w.WriteBatch(ctx, batchToWrite)
 			if perror != nil {
+				// fmt.Printf("DEBUG perror type %s\n", reflect.TypeOf(perror))
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply internal/write/service.go]: removed.

File: internal/write/service.go [hunk 6]
```diff
@@ -196,9 +197,30 @@ func (w *Service) HandleWrite(ctx context.Context, batch *Batch) error {
...
+						logHeaders := perror.HeaderToString([]string{
+							"date",
+							"trace-id",
+							"trace-sampled",
+							"X-Influxdb-Build",
+							"X-Influxdb-Request-ID",
+							"X-Influxdb-Version",
+						})
+						if len(logHeaders) > 0 {
+							logMessage += fmt.Sprintf("\nSelect Response Headers:\n%s", logHeaders)
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply internal/write/service.go]: Changed to "Selected"

File: internal/write/service_test.go (hunk 1 above)
Comment (karel-rehor) [reply internal/write/service_test.go]: removed

File: internal/write/service_test.go (hunk 2 above)
Comment (karel-rehor) [reply internal/write/service_test.go]: removed

File: internal/write/service.go (hunk 3 above)
Comment (karel-rehor) [reply internal/write/service.go]: Agree.  When I wrote it, I was thinking there has to be a more elegant way of implementing this.  I'm still fairly new to golang and the suggested solution did not occur to me.  Implemented as suggested, and with helper methods for extracting header values.

File: internal/write/service.go [hunk 7]
```diff
@@ -196,9 +196,21 @@ func (w *Service) HandleWrite(ctx context.Context, batch *Batch) error {
...
+							"X-Influxdb-Request-ID",
+							"X-Influxdb-Version",
+						})
+						if len(logHeaders) > 0 {
+							logMessage += fmt.Sprintf("\nSelected Response Headers:\n%s", logHeaders)
+						}
+						log.Error(logMessage)
 					}
-					return fmt.Errorf("write failed (attempts %d): %w", batchToWrite.RetryAttempts, perror)
+					return write.NewError(perror, fmt.Sprintf("write failed (retry attempts %d)", batchToWrite.RetryAttempts))
```
Comment (vlastahajek) [inline internal/write/service.go:213]: I'm sorry for the previous suggestion which now looks misleading. I looked at WriteAPI and WriteBlockingAPI and also at several issues regarding complaints about errors. The best solution will be to return `perror`, and log a message about attempts.

**Review by vlastahajek — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,4 +1,8 @@
-## 2.14.0 [Unreleased]
+## 2.14 [unreleased]
+
+### Features
+
+- [#404](https://github.com/influxdata/influxdb-client-go/pull/404) Expose HTTP response headers in the Error type to aid analysis and debugging of error results.  Add selected response headers to the error log.
```
Comment (vlastahajek) [inline CHANGELOG.md:5]:
```suggestion
- [#404](https://github.com/influxdata/influxdb-client-go/pull/404) Expose HTTP response headers in the Error type to aid analysis and debugging of error results.  Add selected response headers to the error log.
Also, unified errors returned by WriteAPI, which now always returns `http.Error`
```

File: api/http/error_test.go [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+	for _, test := range tests {
+		err := Error{
+			StatusCode: test.statusCode,
+			Code:       test.code,
+			Message:    test.message,
+			Err:        test.err,
+			RetryAfter: 0,
+			Header:     ihttp.Header{},
+		}
+		assert.Equal(t, test.expected, err.Error())
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply api/http/error_test.go]: refactored using `t.Run()`

File: api/write/error.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+package write
+
+import (
+	"errors"
+	"fmt"
+	"github.com/influxdata/influxdb-client-go/v2/api/http"
+	iHttp "net/http"
+	"net/textproto"
+)
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply api/write/error.go]: file removed

File: api/write/error_test.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+package write
+
+import (
+	"errors"
+	"fmt"
+	"github.com/influxdata/influxdb-client-go/v2/api/http"
+	"github.com/stretchr/testify/assert"
+	ihttp "net/http"
+	"testing"
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply api/write/error_test.go]: file removed

File: api/write_test.go [hunk 12]
```diff
@@ -8,7 +8,10 @@ import (
 	"fmt"
 	"io"
 	"math"
+	ihttp "net/http"
+	"net/http/httptest"
 	"runtime"
+	"strconv"
 	"strings"
 	"sync"
 	"testing"
```
[the remark this answers is not in the document]
Comment (karel-rehor) [reply api/write_test.go:17]: activated `goimports` in IDE.  Updated.

File: internal/write/service.go (hunk 7 above)
Comment (karel-rehor) [reply internal/write/service.go]: now logs failed attempt count and status code before returning `perror` directly.

**Review by vlastahajek — APPROVED** [review]
Nice improvement 👍 
Good job!

**Pull request merged**